library/bignum_mod_raw.c

```diff
@@ -124,6 +124,34 @@ void mbedtls_mpi_mod_raw_sub( mbedtls_mpi_uint *X,
 
 /* BEGIN MERGE SLOT 3 */
 
+size_t mbedtls_mpi_mod_raw_inv_prime_working_limbs( size_t AN_limbs )
+{
+    /* mbedtls_mpi_mod_raw_inv_prime() needs a temporary for the exponent,
+     * which will be the same size as the modulus and input (AN_limbs),
+     * and additional space to pass to mbedtls_mpi_core_exp_mod(). */
+    return( AN_limbs +
+            mbedtls_mpi_core_exp_mod_working_limbs( AN_limbs, AN_limbs ) );
+}
+
+void mbedtls_mpi_mod_raw_inv_prime( mbedtls_mpi_uint *X,
+                                    const mbedtls_mpi_uint *A,
+                                    const mbedtls_mpi_uint *N,
+                                    size_t AN_limbs,
+                                    const mbedtls_mpi_uint *RR,
+                                    mbedtls_mpi_uint *T )
+{
+    /* Inversion by power: g^|G| = 1 => g^(-1) = g^(|G|-1), and
+     *                       |G| = N - 1, so we want
+     *                 g^(|G|-1) = g^(N - 2)
+     */
+    mbedtls_mpi_uint *Nminus2 = T;
+    (void) mbedtls_mpi_core_sub_int( Nminus2, N, 2, AN_limbs );
+
+    mbedtls_mpi_core_exp_mod( X,
+                              A, N, AN_limbs, Nminus2, AN_limbs,
+                              RR, T + AN_limbs );
```

Comment:
Err, it is documented, in the function above

```
size_t mbedtls_mpi_mod_raw_inv_prime_working_limbs( size_t AN_limbs )
{
    /* mbedtls_mpi_mod_raw_inv_prime() needs a temporary for the exponent,
     * which will be the same size as the modulus and input (AN_limbs),
     * and additional space to pass to mbedtls_mpi_core_exp_mod(). */
    return( AN_limbs +
            mbedtls_mpi_core_exp_mod_working_limbs( AN_limbs, AN_limbs ) );
}
```

These two functions are tightly-coupled, which is a nuisance, but unavoidable (unless you go the PKCS#11 route of calling a function with a NULL parameter to get a "space needed" output; but that is very unwieldy, and tends to produce hard-to-read code).

So: `mbedtls_mpi_mod_raw_inv_prime()` needs some space for itself, and some space to pass on to `mbedtls_mpi_core_exp_mod()`. How much is documented in `mbedtls_mpi_mod_raw_inv_prime_working_limbs()`. Unless `mbedtls_mpi_mod_raw_inv_prime()` is going to call `mbedtls_mpi_core_exp_mod_working_limbs()` itself, to find out how much space that needs, it knows that `mbedtls_mpi_mod_raw_inv_prime_working_limbs()` has given it that amount (for exp_mod) plus the space for a single MPI of AN_limbs (the exponent), so inv_prime uses the first AN_limbs for itself and passes the rest on to exp_mod.

And yes, the temporary space is partitioned exactly as you say. If you feel it's necessary to add this I will, but it feels like duplication.

> We do we not ask the caller to use mbedtls_mpi_mod_raw_inv_prime_working_limbs to allocate the appropriate size of T, and then pass it on as it is?

Yes, that's exactly what we do.  What we don't do is ask the caller to call `mbedtls_mpi_core_exp_mod_working_limbs()`, since then the caller would need to know how `mbedtls_mpi_mod_raw_inv_prime()` is implemented.

Reply:
Following off-line discussion, I've added a couple of comments in `mbedtls_mpi_mod_raw_inv_prime()`